File: csharp/808.soup-servings.cs
/*
 * @lc app=leetcode id=808 lang=csharp
 *
 * [808] Soup Servings
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {

    private static double SoupServingsRecurse(int x, int y, IDictionary<(int X, int Y), double> memo) {
        if (x <= 0) {
            return y <= 0 ? 0.5 : 1.0;
        }
        if (y <= 0) {
            return 0.0;
        }
        var key = (X: x, Y: y);
        double val;
        if (memo.TryGetValue(key, out val)) {
            return val;
        }
        val = (SoupServingsRecurse(x - 100, y, memo) + SoupServingsRecurse(x - 75, y - 25, memo) + SoupServingsRecurse(x - 50, y - 50, memo) + SoupServingsRecurse(x - 25, y - 75, memo)) / 4.0;
        memo[key] = val;
        return val;
    }

    public double SoupServings(int N) {
        return N >= 4500 ? 1.0 : SoupServingsRecurse(N, N, new Dictionary<(int X, int Y), double>());
    }
}
// @lc code=end

File: java/808.soup-servings.java
/*
 * @lc app=leetcode id=808 lang=java
 *
 * [808] Soup Servings
 */

// @lc code=start
class Solution {

    private static double soupServings(int a, int b, double[][] memo) {
        if (a <= 0) {
            return b <= 0 ? 0.5 : 1.0;
        }
        if (b <= 0) {
            return 0.0;
        }
        if (memo[a][b] != 0) {
            return memo[a][b];
        }
        return memo[a][b] = 0.25 * (soupServings(a - 4, b, memo) + soupServings(a - 3, b - 1, memo) + soupServings(a - 2, b - 2, memo) + soupServings(a - 1, b - 3, memo));
    }

    public double soupServings(int N) {
        if (N >= 4500) {
            return 1.0;
        }
        return soupServings((N + 24) / 25, (N + 24) / 25, new double[180][180]);
    }
}
// @lc code=end

